### PR TITLE
Phase06

### PR DESCRIPTION
**Pull request by ali-bonaftizadeh**

### CODE REVIEW

File: phase06/SampleLibrary/SampleLibrary.Test/InvertedIndexTest.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        invertedIndex.indexDocument("hello ali", "1");
+        invertedIndex.indexDocument("this is a code", "2");
+        invertedIndex.indexDocument("to search some words", "3");
+        invertedIndex.indexDocument("in a text, ali", "4");
+
+        var map = invertedIndex.Map;
+
+        map.ContainsKey("hello").Should().Be(true);
+        map.ContainsKey("mohammad").Should().NotBe(true);
```
Comment (aryajalali) [inline phase06/SampleLibrary/SampleLibrary.Test/InvertedIndexTest.cs:20]: Use collection.Should().NotContain or collection.Should().Contain

File: phase06/SampleLibrary/SampleLibrary.Test/InvertedIndexTest.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+public class InvertedIndexTest
+{
+    [Fact]
+    public void InvertedIndex_InsertSomeWords_ReturnsTrue()
+    {
+        var invertedIndex = new InvertedIndex();
+
+        invertedIndex.indexDocument("hello ali", "1");
+        invertedIndex.indexDocument("this is a code", "2");
+        invertedIndex.indexDocument("to search some words", "3");
```
Comment (aryajalali) [inline phase06/SampleLibrary/SampleLibrary.Test/InvertedIndexTest.cs:14]: Initialize somewhere else, and use InlineData.

**Review by aryajalali — CHANGES_REQUESTED** [review]
Add code coverage as well.

File: phase06/ConsoleApp/SearchEngine/RunProgram.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        var io = new IOWorks();
+        var input = io.GetInput();
+
+        var inputProcessor = new InputProcessor();
+        inputProcessor.Process(input);
+
+        var searchInDocs = new SearchInDocs(completedInvertedIndex.Map);
+        HashSet<string> result = searchInDocs.search(inputProcessor.necessary,
+            inputProcessor.optional, inputProcessor.avoided);
+
```
Comment (mojo-izadi) [inline phase06/ConsoleApp/SearchEngine/RunProgram.cs:23]: why do you have three separate inputs for this function?
it is more readable if you just pass 'inputProcessor' to it as a single input.

**Review by mojo-izadi — COMMENTED** [review]
very good code overall
I could only find one thing to comment on

File: phase06/ConsoleApp/SearchEngine/RunProgram.cs (hunk 3 above)
Comment (ali-bonaftizadeh) [reply phase06/ConsoleApp/SearchEngine/RunProgram.cs:23]: You're right, but the problem is with the library and should be corrected from the previous phase code.

Comment (mojo-izadi) [reply phase06/ConsoleApp/SearchEngine/RunProgram.cs:23]: :(

**Review by mojo-izadi — APPROVED** [review]

File: phase06/ConsoleApp/SearchEngine/AppMain.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+using SampleLibrary;
+
+namespace SearchEngine;
+
+public class AppMain
+{
+    public static void Main(string[] args)
+    {
+        var filesDir = @"SearchEngine/EnglishData";
```
Comment (Soltaniant) [inline phase06/ConsoleApp/SearchEngine/AppMain.cs]: you can also use a dedicated file called `config` file for such variables

File: phase06/ConsoleApp/SearchEngine/AppMain.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+        var fileReader = new FileReader(filesDir);
+        new RunProgram().Run(fileReader);
```
Comment (Soltaniant) [inline phase06/ConsoleApp/SearchEngine/AppMain.cs:13]: A program does not necessarily need a `fileReader`. this fact would remind us that we might violate some principles.

File: phase06/ConsoleApp/SearchEngine/IOWorks.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿namespace SearchEngine;
+
+public class IOWorks
+{
+    public String[] GetInput()
```
Comment (Soltaniant) [inline phase06/ConsoleApp/SearchEngine/IOWorks.cs]: Same as what was said for EFCore reviews, our IO is not limited to command line interface or console. its better to use some interfaces
**+** splitting input violates SRP.

File: phase06/ConsoleApp/SearchEngine/IOWorks.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        string input = Console.ReadLine();
+        return input.ToUpper().Split(" ");
+    }
+
+    public void Print(HashSet<string> set) {
```
Comment (Soltaniant) [inline phase06/ConsoleApp/SearchEngine/IOWorks.cs]: getting `IEnumerable` as argument is more general.

File: phase06/ConsoleApp/SearchEngine/AppMain.cs (hunk 4 above)
Comment (matinmoradi80) [reply phase06/ConsoleApp/SearchEngine/AppMain.cs]: جالب بود. ممنون

File: phase06/ConsoleApp/SearchEngine/AppMain.cs (hunk 5 above)
Comment (matinmoradi80) [reply phase06/ConsoleApp/SearchEngine/AppMain.cs:13]: اگر منظور، new کردن کلاس از بیرون و بعد پاس دادنش به تابع Run هست، به نظر میرسد این روش نسبت به انجام این کار در درون تابع Run منطقی تر باشد چراکه از dependency injection استفاده میکنیم و اگر بعدا خواستیم مثلا رفتار آن را تغییر دهیم مثلا در تست تابع Run، راحت تر خواهیم بود.

**Review by amirarab888 — APPROVED** [review]

**Pull request merged**